### ISSUE TITLE
Insert/Update methods do not work for boolean fields in QBMapper

### ISSUE THREAD

**Issue by marius-wieschollek** [nextcloud/server#11236]
I wanted to migrate my app to the new database layer offered by Nextcloud 14. I have several boolean fields declared in the [database.xml](https://github.com/marius-wieschollek/passwords/blob/2019.1.0/src/appinfo/database.xml#L61). After migrating to the new [`insert`](https://github.com/nextcloud/server/blob/stable14/lib/public/AppFramework/Db/QBMapper.php#L102) and [`update`](https://github.com/nextcloud/server/blob/stable14/lib/public/AppFramework/Db/QBMapper.php#L136) methods of the [QBMapper class](https://github.com/nextcloud/server/blob/stable14/lib/public/AppFramework/Db/QBMapper.php), both methods threw an Exception. The Exception (as seen below) suggests that an empty value is given to the database server when the boolean field is set to `false`. I looked in the Nextcloud core code but was unable to find any boolean fields there. After overwriting the insert and update methods and [adding the `$type` parameter for boolean fields](https://github.com/marius-wieschollek/passwords/blob/2019.1.0/src/lib/Db/AbstractMapper.php#L230) the methods work as expected.

### Steps to reproduce
1. Create an app for Nextcloud 14
2. Create a table with a field with the type "boolean" in your database.xml
3. Create a new entry or update an entry for the table wetting the boolean field to `false`
4. Try to use `\OCP\AppFramework\Db\QBMapper::insert` or `\OCP\AppFramework\Db\QBMapper::update` to insert or update an entry for the table from 2. 

### Expected behaviour
Entry should be created or updated without issue.

### Actual behaviour
An Exception is thrown telling you that you tried to set an invalid value for the boolean field:
`Doctrine\DBAL\Driver\PDOException: SQLSTATE[22007]: Invalid datetime format: 1366 Incorrect integer value: '' for column 'booleancolumn' at row 1`

### Server configuration
**Operating system**: Ubuntu 16.04
**Web server:** Nginx
**Database:** MariaDB 10
**PHP version:** 7.2.10
**Nextcloud version:** 14.0.0.19
**Updated from an older Nextcloud/ownCloud or fresh install:** Upgrade from 13.0.6
**Where did you install Nextcloud from:** Official Docker Image

**Signing status:**
<details>
<summary>Signing status</summary>

```
No errors have been found.
```
</details>

**List of activated apps:**
<details>
<summary>App list</summary>

```
Enabled:
  - accessibility: 1.0.1
  - activity: 2.7.0
  - cloud_federation_api: 0.0.1
  - comments: 1.4.0
  - dav: 1.6.0
  - federatedfilesharing: 1.4.0
  - federation: 1.4.0
  - files: 1.9.0
  - files_pdfviewer: 1.3.2
  - files_sharing: 1.6.2
  - files_texteditor: 2.6.0
  - files_trashbin: 1.4.1
  - files_versions: 1.7.1
  - files_videoplayer: 1.3.0
  - firstrunwizard: 2.3.0
  - gallery: 18.1.0
  - logreader: 2.0.0
  - lookup_server_connector: 1.2.0
  - nextcloud_announcements: 1.3.0
  - notifications: 2.2.1
  - oauth2: 1.2.1
  - password_policy: 1.4.0
  - passwords: 2019.1.0-BUILD
  - provisioning_api: 1.4.0
  - serverinfo: 1.4.0
  - sharebymail: 1.4.0
  - support: 1.0.0
  - survey_client: 1.2.0
  - systemtags: 1.4.0
  - theming: 1.5.0
  - twofactor_backupcodes: 1.3.1
  - updatenotification: 1.4.1
  - workflowengine: 1.4.0
Disabled:
  - admin_audit
  - encryption
  - files_external
  - user_external
  - user_ldap
```
</details>

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
{
    "system": {
        "memcache.local": "\\OC\\Memcache\\APCu",
        "apps_paths": [
            {
                "path": "\/var\/www\/html\/apps",
                "url": "\/apps",
                "writable": false
            },
            {
                "path": "\/var\/www\/html\/custom_apps",
                "url": "\/custom_apps",
                "writable": true
            }
        ],
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "localhost"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "14.0.0.19",
        "overwrite.cli.url": "https:\/\/localhost",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "loglevel": 0,
        "theme": "",
        "maintenance": false
    }
}
```
</details>

**Are you using external storage, if yes which one:** no
**Are you using encryption:** no
**Are you using an external user-backend, if yes which one:** no

### Client configuration
**Browser:** Firefox 62
**Operating system:** Ubuntu 16.04

#### Nextcloud log (data/nextcloud.log)
<details>
<summary>Nextcloud log</summary>

```
Doctrine\DBAL\Exception\DriverException: An exception occurred while executing 'INSERT INTO `passwords_entity_folder_revision` (`user_id`, `uuid`, `deleted`, `created`, `updated`, `model`, `favorite`, `label`, `parent`, `cse_type`, `hidden`, `trashed`, `edited`, `sse_type`, `client`, `sse_key`) VALUES(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)' with params ["admin", "671e4c75-9f51-491d-87dd-3f571414a6c1", false, 1537039395, 1537039395, "e4af1881-7fea-442f-8ea2-ac89e65b42f7", false, "replaced-nonessential-value", "00000000-0000-0000-0000-000000000000", "none", false, false, 1537039395, "SSEv1r1", "", "replaced-nonessential-value"]: SQLSTATE[22007]: Invalid datetime format: 1366 Incorrect integer value: '' for column 'deleted' at row 1
    /var/www/html/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php - line 128:
    Doctrine\DBAL\Driver\AbstractMySQLDriver->convertException("An exceptio ... 1", Doctrine\DBA ... "})
    /var/www/html/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Connection.php - line 1015:
    Doctrine\DBAL\DBALException::driverExceptionDuringQuery(Doctrine\DBA ... {}, Doctrine\DBA ... "}, "INSERT INTO ... )", { 1: "admin" ... "})
    /var/www/html/lib/private/DB/Connection.php - line 216:
    Doctrine\DBAL\Connection->executeUpdate("INSERT INTO ... )", [ "admin","6 ... "], [ 2,2,2,2,2, ... 2])
    /var/www/html/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Query/QueryBuilder.php - line 208:
    OC\DB\Connection->executeUpdate("INSERT INTO ... )", { dcValue1: ... "}, { dcValue1: ... 2})
    /var/www/html/lib/private/DB/QueryBuilder/QueryBuilder.php - line 214:
    Doctrine\DBAL\Query\QueryBuilder->execute()
    /var/www/html/lib/public/AppFramework/Db/QBMapper.php - line 119:
    OC\DB\QueryBuilder\QueryBuilder->execute()
    /var/www/html/custom_apps/passwords/lib/Services/Object/AbstractRevisionService.php - line 147:
    OCP\AppFramework\Db\QBMapper->insert(OCA\Password ... l})
    /var/www/html/custom_apps/passwords/lib/Services/Object/FolderRevisionService.php - line 143:
    OCA\Passwords\Services\Object\AbstractRevisionService->save(OCA\Password ... l})
    /var/www/html/custom_apps/passwords/lib/Controller/Api/FolderApiController.php - line 98:
    OCA\Passwords\Services\Object\FolderRevisionService->save(OCA\Password ... l})
    /var/www/html/lib/private/AppFramework/Http/Dispatcher.php - line 166:
    OCA\Passwords\Controller\Api\FolderApiController->create("asdfdfas", "00000000-00 ... 0", "none", 1537039395, false, false)
    /var/www/html/lib/private/AppFramework/Http/Dispatcher.php - line 99:
    OC\AppFramework\Http\Dispatcher->executeController(OCA\Password ... {}, "create")
    /var/www/html/lib/private/AppFramework/App.php - line 118:
    OC\AppFramework\Http\Dispatcher->dispatch(OCA\Password ... {}, "create")
    /var/www/html/lib/private/AppFramework/Routing/RouteActionHandler.php - line 47:
    OC\AppFramework\App::main("FolderApiController", "create", OC\AppFramew ... {}, { _route: "p ... "})
    OC\AppFramework\Routing\RouteActionHandler->__invoke({ _route: "p ... "})
    /var/www/html/lib/private/Route/Router.php - line 297:
    call_user_func(OC\AppFramew ... {}, { _route: "p ... "})
    /var/www/html/lib/base.php - line 989:
    OC\Route\Router->match("/apps/passw ... e")
    /var/www/html/index.php - line 42:
    OC::handleRequest()
	
Eingefasst vonDoctrine\DBAL\Driver\PDOException: SQLSTATE[22007]: Invalid datetime format: 1366 Incorrect integer value: '' for column 'deleted' at row 1
    /var/www/html/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Connection.php - line 1006:
    Doctrine\DBAL\Driver\PDOStatement->execute()
    /var/www/html/lib/private/DB/Connection.php - line 216:
    Doctrine\DBAL\Connection->executeUpdate("INSERT INTO ... )", [ "admin","6 ... "], [ 2,2,2,2,2, ... 2])
    /var/www/html/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Query/QueryBuilder.php - line 208:
    OC\DB\Connection->executeUpdate("INSERT INTO ... )", { dcValue1: ... "}, { dcValue1: ... 2})
    /var/www/html/lib/private/DB/QueryBuilder/QueryBuilder.php - line 214:
    Doctrine\DBAL\Query\QueryBuilder->execute()
    /var/www/html/lib/public/AppFramework/Db/QBMapper.php - line 119:
    OC\DB\QueryBuilder\QueryBuilder->execute()
    /var/www/html/custom_apps/passwords/lib/Services/Object/AbstractRevisionService.php - line 147:
    OCP\AppFramework\Db\QBMapper->insert(OCA\Password ... l})
    /var/www/html/custom_apps/passwords/lib/Services/Object/FolderRevisionService.php - line 143:
    OCA\Passwords\Services\Object\AbstractRevisionService->save(OCA\Password ... l})
    /var/www/html/custom_apps/passwords/lib/Controller/Api/FolderApiController.php - line 98:
    OCA\Passwords\Services\Object\FolderRevisionService->save(OCA\Password ... l})
    /var/www/html/lib/private/AppFramework/Http/Dispatcher.php - line 166:
    OCA\Passwords\Controller\Api\FolderApiController->create("asdfdfas", "00000000-00 ... 0", "none", 1537039395, false, false)
    /var/www/html/lib/private/AppFramework/Http/Dispatcher.php - line 99:
    OC\AppFramework\Http\Dispatcher->executeController(OCA\Password ... {}, "create")
    /var/www/html/lib/private/AppFramework/App.php - line 118:
    OC\AppFramework\Http\Dispatcher->dispatch(OCA\Password ... {}, "create")
    /var/www/html/lib/private/AppFramework/Routing/RouteActionHandler.php - line 47:
    OC\AppFramework\App::main("FolderApiController", "create", OC\AppFramew ... {}, { _route: "p ... "})
    OC\AppFramework\Routing\RouteActionHandler->__invoke({ _route: "p ... "})
    /var/www/html/lib/private/Route/Router.php - line 297:
    call_user_func(OC\AppFramew ... {}, { _route: "p ... "})
    /var/www/html/lib/base.php - line 989:
    OC\Route\Router->match("/apps/passw ... e")
    /var/www/html/index.php - line 42:
    OC::handleRequest()
	
Eingefasst vonPDOException: SQLSTATE[22007]: Invalid datetime format: 1366 Incorrect integer value: '' for column 'deleted' at row 1
    /var/www/html/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php - line 105:
    PDOStatement->execute(null)
    /var/www/html/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Connection.php - line 1006:
    Doctrine\DBAL\Driver\PDOStatement->execute()
    /var/www/html/lib/private/DB/Connection.php - line 216:
    Doctrine\DBAL\Connection->executeUpdate("INSERT INTO ... )", [ "admin","6 ... "], [ 2,2,2,2,2, ... 2])
    /var/www/html/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Query/QueryBuilder.php - line 208:
    OC\DB\Connection->executeUpdate("INSERT INTO ... )", { dcValue1: ... "}, { dcValue1: ... 2})
    /var/www/html/lib/private/DB/QueryBuilder/QueryBuilder.php - line 214:
    Doctrine\DBAL\Query\QueryBuilder->execute()
    /var/www/html/lib/public/AppFramework/Db/QBMapper.php - line 119:
    OC\DB\QueryBuilder\QueryBuilder->execute()
    /var/www/html/custom_apps/passwords/lib/Services/Object/AbstractRevisionService.php - line 147:
    OCP\AppFramework\Db\QBMapper->insert(OCA\Password ... l})
    /var/www/html/custom_apps/passwords/lib/Services/Object/FolderRevisionService.php - line 143:
    OCA\Passwords\Services\Object\AbstractRevisionService->save(OCA\Password ... l})
    /var/www/html/custom_apps/passwords/lib/Controller/Api/FolderApiController.php - line 98:
    OCA\Passwords\Services\Object\FolderRevisionService->save(OCA\Password ... l})
    /var/www/html/lib/private/AppFramework/Http/Dispatcher.php - line 166:
    OCA\Passwords\Controller\Api\FolderApiController->create("asdfdfas", "00000000-00 ... 0", "none", 1537039395, false, false)
    /var/www/html/lib/private/AppFramework/Http/Dispatcher.php - line 99:
    OC\AppFramework\Http\Dispatcher->executeController(OCA\Password ... {}, "create")
    /var/www/html/lib/private/AppFramework/App.php - line 118:
    OC\AppFramework\Http\Dispatcher->dispatch(OCA\Password ... {}, "create")
    /var/www/html/lib/private/AppFramework/Routing/RouteActionHandler.php - line 47:
    OC\AppFramework\App::main("FolderApiController", "create", OC\AppFramew ... {}, { _route: "p ... "})
    OC\AppFramework\Routing\RouteActionHandler->__invoke({ _route: "p ... "})
    /var/www/html/lib/private/Route/Router.php - line 297:
    call_user_func(OC\AppFramew ... {}, { _route: "p ... "})
    /var/www/html/lib/base.php - line 989:
    OC\Route\Router->match("/apps/passw ... e")
    /var/www/html/index.php - line 42:
    OC::handleRequest()
```
</details>

**Comment by MorrisJobke** [comment]
cc @nickvergessen

**Comment by ChristophWurst** [comment]
@rullzer is that why you always register the attribute types like in https://github.com/nextcloud/server/blob/240d919cefc3b9ca1d94fc10bdda604defe99c81/lib/private/Authentication/Token/DefaultToken.php#L93?

**Comment by rullzer** [comment]
Yes well I like to be explicit. But marking them as 'bool' there should do the trick i think indeed.

**Comment by marius-wieschollek** [comment]
I have had a patch for this in my app since December and i did not see any errors with it since then.
So if it's ok with you, i would like to create a pr to fix that issue permanently.

**Comment by MorrisJobke** [comment]
Fix in #14825
